orttraining/orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cpu/optimizer/clip_grad_norm/clip_grad_norm.h"
+#include "core/providers/cpu/math/element_wise_ops.h"
+#include "core/providers/cpu/tensor/utils.h"
+#include "core/providers/cpu/reduction/reduction_ops.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+namespace {
+
+constexpr float Epsilon = 0.000001f;
+
+template <typename T>
+T GetL2Norm(const TensorSeq& gradients) {
+  T l2_norm = 0;
+  for (const auto& tensor : gradients) {
+    l2_norm +=
+        ReduceAggregatorSumSquare<T>(tensor.Shape().Size(), *tensor.Data<T>()).aggall(tensor.Data<T>());
+  }
+  return reduce_sqrt<T>(l2_norm);
+}
+
+template <typename T>
+void ClipGradNorm(T total_norm, T max_norm, TensorSeq& gradients) {
+  const T clip_coefficient = std::min(max_norm / (total_norm + static_cast<T>(Epsilon)), static_cast<T>(1.0f));
+
+  for (const auto& grad : gradients) {
+    auto& tensor = const_cast<Tensor&>(grad);
```

Comment:
would it be possible to update TensorSeq to allow mutable access to its tensors?

Reply:
So far, we have 2 ops in training that would benefit from this (AdamWOptimizer and InplaceClipGradNorm).
I think we probably should. I haven't taken a look at other examples yet, but I am wondering how one would go about performing operations on the tensors inside the tensorseq (other than recreate new tensors from existing ones) if the only access is non mutable.
Having said that, I think I would like to probably get to that in another PR. I can create a work item for this.